keps/sig-cloud-provider/1959-service-lb-class-field/README.md

```diff
@@ -0,0 +1,217 @@
+* allow multiple implementations of Service Type=LoadBalancer in a cluster.
+* prevent every cloud provider from implementing a custom "opt-out" annotation for their load balancer.
+
+### Non-Goals
+
+* performance improvements for Service Type=LoadBalancer.
+* changing any other existing behaviors for Service Type=LoadBalancer aside from being able
+to disabling it from the cloud provider.
+
+## Proposal
+
+This KEP proposes to add a new field `spec.loadBalancerClass` in Service which allows for
+multiple implementations of Service Type=LoadBalancer in a cluster.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin:
+* I want to use my cloud provider's public load balancer service for applications that require
+public ingress.
+* I want to use my own load balancing solution for any applications that only talk internally
+within my own network because I want to save costs.
+
+#### Story 2
+
+As an application developer:
+* I MUST use a hardware-based loadbalancer for certain applications due to specific protocols
+only available there.
+* I want to use the cloud provider's default load balancer for any applications that do not
+rely on protocols from hardware load balancers.
+
+### Risks and Mitigations
+
+Many cloud providers today support an "opt-out" annotation for this behavior. The annotation is specific
+to the cloud provider. Introduction of the `loadBalancerClass` field at this point would mean that
+cloud providers need to start accounting for both existing annotations and the new field.
+
+## Design Details
+
+Introduce a new field to Service `spec.loadBalancerClass`.
+
+If the field `spec.loadBalancerClass` is not set, the existing cloud provider will assume
+ownership of the Service Type=LoadBalancer resource. This is required to not break existing clusters
+that assume Service Type=LoadBalancer is always managed by the cloud provider.
+
+Required API changes:
+```go
+// ServiceSpec describes the attributes that a user creates on a service.
+type ServiceSpec struct {
+	...
+	...
+
+	// loadBalancerClass is the name of the load balancer implementation this Service belongs to.
+	// This field can only be set when the Service type is 'LoadBalancer'. If not set, the default load
+	// balancer implementation is used, today this is typically done through the cloud provider integration,
+	// but should apply for any default implementation. If set, it is assumed that a load balancer
+	// implementation is watching for Services with a matching class name. Any default load balancer
+	// implementation (e.g. cloud providers) should ignore Services that set this field.
+	// +optional
+	LoadBalancerClassName string `json:"loadBalancerClass,omitempty"`
+}
+```
+
+* `loadBalancerClass` will be immutable only when the Service type is `LoadBalancer`, this way existing and future implementations
+do not have to worry about handling Services that change the class name. The class name is mutable and can be cleared when the
+type changes.
```

Comment:
Need to add a section for Garbage Collection.

It needs to be compatible with the current finalizer implementation.